### PR TITLE
feat(model-lab): reproducible fine-tuning lab — faithfulness-gate + correction-intent (#1585)

### LINKED ISSUE #1585: model-lab: RTX 3090 fine-tuning lab — faithfulness-gate and correction-intent models
Part of #1572. Standards: #1520. Feeds: #1576 (gate backend), #1581 (intent classifier). Depends on: #1576 shadow mode (training-data harvest), #1584 (downstream eval).

## Problem

Two new classification workloads — the extraction **faithfulness gate** and **correction-intent detection** — will run on every extraction flush. Prompted frontier LLMs work but are slow, non-deterministic, rate-limited, and costly at that call frequency. Both tasks are narrow classification/extraction problems that small fine-tuned models excel at, and the lab hardware (RTX 3090 24 GB, 256 GB RAM, NVMe) is exactly the right size for them.

This issue creates the **model lab**: reproducible recipes to generate training data, fine-tune, evaluate, and serve two small models that Remnic's config can point at — with everything reproducible from manifests and **no datasets or weights ever committed to this repo**.

## Hardware reality (sets the model-size policy)

| Operation | RTX 3090 (24 GB, Ampere — bf16 yes, FP8 no) |
|---|---|
| Encoder classifier fine-tune (DeBERTa-v3-large class, ~0.4B) | Trivial: full fine-tune, minutes–1h |
| Full fine-tune, causal LM | ≤ ~1.5B comfortably |
| LoRA (bf16 base) | ≤ ~8B |
| QLoRA (4-bit base + paged optimizer) | ≤ ~14B (bs 1–2 + grad-accum); ~30B-class technically possible but too slow to iterate — out of policy |
| Inference serving (Ollama / vLLM, 4-bit) | ~14B dense or ~30B-class MoE, one model at a time |

**Policy: target models ≤ 4B for these two tasks** (they are classification, not generation — small models + good data win), with an 8B LoRA escape hatch if evals demand it. Anything larger needs a written justification in the manifest.

### Layout — top-level `model-lab/` (scripts + manifests, NOT an npm workspace package)

```
model-lab/
  README.md                     # hardware envelope, policy table above, quickstart
  common/                       # shared: seeding, JSONL schema, eval runners, HF push helper
  faithfulness-gate/
    generate-data.(ts|py)       # synthetic + harvest pipelines (below)
    train.py                    # single-GPU recipe (TRL/unsloth-class stack)
    eval.py                     # held-out metrics + downstream hooks
    manifest.json               # THE reproducibility artifact (schema below)
  correction-intent/
    (same shape)
.gitignore additions: model-lab/**/data/, model-lab/**/runs/, model-lab/**/*.safetensors, *.gguf
```

Manifest schema (checked into git; the only thing that is): `{task, baseModel: {hfId, revision}, dataRecipe: {generatorGitSha, seed, sources, counts, datasetSha256}, hyperparams, trainedAt, hardware, eval: {heldOut: {...}, downstream: {...}}, artifact: {hfRepo, revision, quantizations}}`. Weights are published to the operator's HF account and referenced by the manifest — git carries recipes and hashes, never blobs (repo ethics rule; checklist §3 — no secrets/PII in any committed file).

### Model 1 — `remnic-faithfulness-gate-v1`

- **Task:** (factText, quote, context) → {entailed, contradicted, unsupported}. Exactly the `FaithfulnessCheckInput` contract from #1576.
- **Base candidates:** (a) DeBERTa-v3-large-MNLI-class encoder with a 3-way head — first choice: cheapest, deterministic, ~1.6 GB serving; (b) a ≤4B instruct causal LM fine-tuned to emit one token class — fallback if the encoder plateaus on long contexts.
- **Training data, three streams:**
  1. **Synthetic perturbation** (the volume driver): take extraction fixtures + generated (quote, fact) pairs; programmatic perturbations create labels — entity swap / negation flip / date shift / quantity change ⇒ `contradicted`; unrelated-quote pairing ⇒ `unsupported`; identity/paraphrase (frontier-model paraphrases of the fact, verified back) ⇒ `entailed`. Perturbations are code, so labels are trustworthy by construction.
  2. **Shadow-log harvest**: #1576 shadow mode records (fact, quote, verdict) with the prompted frontier model as teacher. Harvest is **opt-in, local-only, and documented** — a `model-lab` script reads the operator's own telemetry from their own memoryDir; nothing leaves the machine; the committed dataset hash covers only synthetic + the operator's consented harvest.
  3. **Public NLI warm-start** (optional): standard NLI corpora for pretraining the head before task data.
- **Eval:** held-out F1 per class (target: macro-F1 ≥ 0.9 on synthetic held-out before any downstream claim) + **downstream**: plug into `extraction.faithfulnessModel`, run the #1574 LoCoMo adversarial-category ablation and compare against the prompted-LLM gate on both accuracy and p95 gate latency. The downstream number is the one that matters (rule 55).

### Model 2 — `remnic-correction-intent-v1`

- **Task:** turn text (+ small prior-turn window) → the #1581 `corrections[]` block: {targetHint, correctedAssertion, polarity, confidence} or none. Structured extraction → a ≤4B instruct LM with JSON-schema output is the right shape (an encoder can't emit the spans).
- **Training data:** synthetic conversation generator (personas correcting agents across the full morphology matrix from #1581's fixture taxonomy — reuse those fixtures as the seed grammar) + the anti-example classes (quotes, third-party, hypotheticals) labeled `none`. Frontier teacher labels a sampled slice for validation of the generator's label quality (report agreement in the manifest).
- **Eval:** held-out detection F1 + span quality; **downstream:** MemCorrect (#1584) `false_apply` and `uptake@next` in passive-queue mode vs the prompted classifier.

### Serving + integration

Export merged fp16 + GGUF/AWQ quantizations; serve via Ollama or vLLM on the lab box. Remnic consumes them through existing config only: `extraction.faithfulnessModel` and the #1581 classifier routing pointing at the local openai-compatible endpoint — **zero Remnic-core code changes in this issue**; if a change is needed, it belongs in the consuming child.

## Implementation guide

**PR 1 — lab scaffolding + faithfulness data pipeline.** README (policy table), common utils, gitignore rules, faithfulness `generate-data` with seeded determinism (same seed → same dataset sha256 — test in CI without committing data: generate small-N in a temp dir and assert the hash), perturbation unit tests (each perturbation provably changes/preserves the right label on hand-written cases).

**PR 2 — faithfulness train + eval + manifest.** Train encoder baseline; eval script emits the manifest `eval` block; push to HF; commit manifest. Downstream ablation run + artifact per #1574 protocol.

**PR 3 — correction-intent pipeline + train + eval + manifest.** Same shape; downstream via MemCorrect.

**PR 4 — docs.** `docs/model-lab.md`: how an operator reproduces both models from manifests on their own 3090-class GPU, and how to point Remnic config at them. Every claim traces to a manifest eval block (rule 55).

## Pitfalls

- Python code here follows the global Python standards (3.12+, typed, `requirements.txt`/`setup.sh` local to `model-lab/`) but is NOT part of the npm build/test pipeline — CI must not require a GPU; the only CI hooks are the seeded-generator hash tests (pure CPU, small N).
- Never harvest shadow logs by default; the harvest script requires an explicit `--i-consent-local-data` flag and prints what it will read.
- Teacher-model outputs used for training are "LLM traces" — they live under the gitignored data dir like everything else.
- Version-pin the training stack in the manifest (exact lib versions); a manifest that can't reproduce its own eval numbers is a bug.
- Don't chase 14B models because they fit — iteration speed on a single 3090 is the scarce resource; the policy table is the guardrail.

## Done when

- Both v1 manifests committed with held-out AND downstream eval numbers.
- Remnic on the lab box runs the faithfulness gate + passive correction detection fully locally (no cloud calls) at documented p95 latencies.
- Downstream deltas recorded in bench artifacts (gate quality vs prompted-LLM baseline; false_apply vs prompted baseline).
- Reproduction docs verified by re-running data-gen → hash match.

### PR DESCRIPTION
**Pull request by joshuaswarren**
Closes #1585 (software half). Part of #1572.

## What this lands

The **reproducible software** for the model-lab: recipes, the version-pinned training stack, the manifest schema, both seeded data generators, the eval harness (held-out p95 latency + faithfulness F1 / correction-intent detection F1), and the Remnic config pointers that let the gate + passive-correction detection route to a fine-tuned local model when available. **No GPU is required to land this PR.** Everything is unit-tested on CPU.

Per rule 55 (#1520): **no accuracy/latency/eval number appears anywhere** without a real-run artifact. Both committed manifests ship as `pending-training` schema examples; the actual training runs + eval-number manifests are the **GPU-gated #1585 follow-up** (see *GPU follow-up* below).

### Pieces
- **Manifest schema** (`common/manifest_schema.py`) — the ONLY committed artifact for a trained model. Two first-class states: `allow_pending=True` (committed scaffold, every eval/weight/training-stack field null/pending) and `allow_pending=False` (a real run must pin everything). Proven to reject a manifest missing the `trainingStack` version-pin block (rule 6: prove-fail-before).
- **Training-stack pin** (`common/training_stack.py`) — `requirements_versions` parses the committed `requirements.txt` (exact lib versions); `capture_training_stack` records `pip freeze` at train time (GPU-gated); `assert_stack_matches_requirements` surfaces silent drift between the pin and the run. *A manifest that can't reproduce its own eval numbers is a bug.*
- **Data generators** — faithfulness perturbations (`faithfulness-gate/`) + correction-intent morphology (`correction-intent/`). Seeded determinism: same seed → byte-identical dataset → identical sha256 (asserted in CI). Harvest stream is a consent-gated stub (`--i-consent-local-data`) waiting for #1576/#1581 shadow mode.
- **Eval harness** (`common/eval_runner.py` + `common/latency.py`) — per-class/macro F1 (faithfulness) + detection F1 + span overlap (correction-intent) + held-out p95 latency percentile math. Identical math in CI and on GPU.
- **Config pointers** (`extractionFaithfulnessBaseUrl`, `correctionIntentModel`, `correctionIntentBaseUrl`) — when the gate pointer keys are both set, the gate routes to the local served endpoint first and falls back to the configured chain on any failure (graceful degradation; a down local model never blocks writes). Empty defaults preserve byte-identical pre-feature routing (rule 39). Response narrowing uses in/typeof guards, no inline casts.
- **Docs** (`docs/model-lab.md`, `model-lab/README.md`) — reproducible recipe + hardware envelope + policy table + how to point Remnic config at a served model.

### Wiring fixes (so the committed config pointers pass the gates)
- `validate-config-contract.ts`: teach `getParseConfigReturnKeys` to follow `...parseXxxConfig(cfg)` spreads to the helper's own object-literal return. Supports the #1526 extraction pattern that keeps `config.ts` from growing; without it every extracted key looked "missing" from `parseConfig`'s return. General improvement.
- `openclaw.plugin.json`: declare the 3 new model-lab pointer keys in `configSchema` so the PluginConfig ↔ schema contract holds.
- `extraction-faithfulness.test.ts`: fix `fallbackCalls` array type to match `stubFallbackLlm`'s captured-parameter shape.

## Chokepoints used / not applicable
- **not applicable** — model-lab is a standalone Python tree, not an npm workspace package; it consumes Remnic only through existing config keys. The config pointers are empty-default optional keys (no new `config.*Enabled` reads, no new MCP/HTTP/CLI surfaces).
- ratchet: `config.ts` 4642 → 4621 (extraction improvement; baseline tightened).
- config-contract: validator now spread-aware; schema synchronized.

## Repo ethics
`model-lab/**/data/`, `model-lab/**/runs/`, `model-lab/.venv/`, `*.safetensors`, `*.gguf` are gitignored. The committed manifests are the ONLY model-lab artifacts in git. Verified: no data/weights/LLM-traces tracked.

## Verification (no GPU)
- `npm run preflight:quick` → **OK (quick)** (lint, check-types, check-config-contract, plugin:inspect, check-review-patterns, check-ratchets, check-docs-parity, entity-hardening, quick tests).
- `node scripts/check-ratchets.mjs` → OK.
- Model-lab node:test suite (11 tests, capability-guarded on `python3`): faithfulness + correction-intent determinism/selfcheck, manifest-schema validate + prove-fail-before → all pass.
- Config-pointer unit suite (89 tests): `faithfulness-config`, `local-model-endpoint`, `extraction-faithfulness` → all pass.

## GPU follow-up (NOT in this PR — explicitly gated)
The **actual training runs + committed eval-number manifests + downstream bench artifacts** are GPU-gated. They land in a separate follow-up when the lab box (RTX 3090) frees:
1. Train encoder baseline → fill `faithfulness-gate/manifest.json` `eval.heldOut` + `trainingStack` (`allow_pending=False` must pass).
2. Train ≤4B instruct-LM correction-intent model → fill `correction-intent/manifest.json`.
3. Downstream: #1574 LoCoMo adversarial-category ablation (gate quality vs prompted-LLM baseline, p95 latency); MemCorrect (#1584) `false_apply` / `uptake@next`.
4. Publish weights to operator HF account; reference in manifest `artifact` block.

---

> [!NOTE]
> **Medium Risk**
> Touches extraction faithfulness LLM routing and config parsing; behavior is gated on new optional keys with extensive tests, but misconfiguration or timeout edge cases could affect gate availability until fallback runs.
> 
> **Overview**
> Adds the **CPU-landed software half** of `model-lab/` for faithfulness-gate and correction-intent: manifest schema + `trainingStack` pinning, shared eval/latency math, correction-intent synthetic data + train/eval recipes, and `pending-training` manifests (no fabricated eval numbers).
> 
> **Remnic wiring:** optional config keys `extractionFaithfulnessBaseUrl`, `correctionIntentModel`, and `correctionIntentBaseUrl` (parsed via extracted `faithfulness-config.ts`); the faithfulness gate **prefers** a local OpenAI-compatible endpoint when both URL and model are set, with half-batch probe timeout, fallback to the existing LLM chain, and no leaking of the local model name into fallback on outage. Plugin JSON + spread-aware `validate-config-contract` keep the config contract honest.
> 
> **Tests/docs:** Node probes for correction-intent determinism, manifest strict validation, and eval integrity guards; expanded `docs/model-lab.md` / README. Actual GPU training runs and eval-filled manifests remain explicitly out of scope.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 612c159dcfa450f8416ea0b970b06abf28f1714c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: model-lab/common/manifest_schema.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    # eval + artifact blocks: structural presence only. allow_pending controls
+    # whether the null/pending placeholder is accepted.
+    for block_key in ("eval", "artifact"):
+        block = manifest.get(block_key)
+        if _is_pending(block):
+            if not allow_pending:
+                errors.append(f"{block_key} is pending — a real run must record it")
+        elif not isinstance(block, Mapping):
+            errors.append(f"{block_key} must be an object (or null when pending)")
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/manifest_schema.py:127]: Fixed in a326af22d: strict mode now rejects a trained manifest whose eval object leaves eval.heldOut null (and artifact.hfRepo null). Verified on CPU.

File: model-lab/correction-intent/eval.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            p_assert = str(pred_corr[0].get("correctedAssertion", ""))
+        overlaps.append(span_overlap(p_assert, g_assert))
+    return overlaps
+
+
+def main(argv: list[str] | None = None) -> int:
+    parser = build_arg_parser()
+    args = parser.parse_args(argv)
+
+    require_eval_deps()
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:140]: Not fixing in this PR — intentional, deferring to the GPU-run follow-up. `require_eval_deps()` is the model-lab's invariant: the recipe is designed to run inside the lab venv (`setup.sh` + `.venv`) where torch/transformers are present, and the model-inference loop landing in the #1585 GPU-run follow-up needs them. The offline `--predictions` path runs in that same venv. Relaxing the gate for CPU-only scoring only makes sense once the inference-vs-offline split is real (the follow-up), at which point the gate can be scoped to inference mode. Logging as a follow-up nit on #1585.

File: model-lab/common/training_stack.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    )
+    libs: dict[str, str] = {}
+    for raw in freeze_text.splitlines():
+        line = raw.strip()
+        if "==" not in line:
+            continue
+        name, _, version = line.partition("==")
+        name = name.split("[", 1)[0].strip().lower()
+        if name in PINNED_LIBS:
+            libs[name] = version
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/training_stack.py:113]: Fixed in b61f427a3. `capture_training_stack` and `assert_stack_matches_requirements` now PEP 503-normalize distribution names (underscore → hyphen) before matching, so a `huggingface_hub` freeze line is recorded under the canonical `huggingface-hub` key and no longer silently omitted from `stack.libs` / falsely reported missing. Prove-fail test in `tests/model-lab-integrity-guards.test.mjs` (underscore capture matches a hyphen pin with no error).

File: model-lab/common/manifest_schema.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            errors.append("trainingStack is pending — a real run must pin exact lib versions")
+    elif not isinstance(stack, Mapping):
+        errors.append("trainingStack must be an object (or null when pending)")
+    else:
+        stack_errs = _validate_training_stack(stack, allow_pending=allow_pending)
+        errors.extend(stack_errs)
+
+    # eval + artifact blocks: structural presence only. allow_pending controls
+    # whether the null/pending placeholder is accepted.
+    for block_key in ("eval", "artifact"):
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/manifest_schema.py]: Fixed in b61f427a3. Strict mode (`allow_pending=False`) now requires concrete `dataRecipe` provenance — `generatorGitSha` and `datasetSha256` must both be non-null for a `status:"trained"` manifest, so a published run can't omit the seed/dataset-hash and pass the reproducibility gate. Prove-fail test asserts a trained manifest with `datasetSha256: null` is rejected, and a fully-recorded trained manifest still passes (no false positives).

File: model-lab/correction-intent/eval.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        print(
+            "eval.py: --predictions <model-output.jsonl> is required to score the model.\n"
+            "  Run the trained checkpoint's inference over the held-out split first;\n"
+            "  scoring gold labels against themselves is refused (no fabricated evals, rule 55).",
+            file=sys.stderr,
+        )
+        return 2
+
+    gold_rows = load_jsonl(args.held_out)
+    pred_rows = load_jsonl(args.predictions)
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:179]: Fixed in b61f427a3. `eval.py` now refuses the held-out file (or a symlink/hardlink to it) as `--predictions` via `Path.resolve()` comparison before loading rows — scoring gold against itself is rejected with exit 2 (rule 55). The guard lives in a pure, torch-free helper `_held_out_is_predictions`, unit-tested for same-file / distinct-file / symlink. This is the check the inline comment (L129) described but the prior revision never wired in.

File: packages/remnic-core/src/extraction-faithfulness.ts [hunk 6]
```diff
@@ -280,12 +284,44 @@ async function callFaithfulnessLlm(
...
+      { temperature: 0.1, maxTokens: 2048, responseFormatJson: true, timeoutMs: probeBudgetMs },
+      fetchImpl,
+    );
+    if (result?.content) {
+      return { content: result.content, modelUsed: result.modelUsed };
+    }
+    log.debug(
+      "extraction-faithfulness: local model-lab endpoint unavailable, trying configured chain",
+    );
+  }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/extraction-faithfulness.ts:323]: Fixed in b61f427a3. When a local model-lab endpoint is configured, `extractionFaithfulnessModel` is now scoped to that endpoint only — it no longer leaks into the fallback chain as a gateway/local-LLM `modelOverride` on endpoint failure, so a local outage degrades gracefully to the configured chain instead of `backend_unavailable`. The pre-feature model-override contract is preserved byte-identical when no endpoint is set (rule 39; the existing override-routes-to-gateway test still passes). Prove-fail test asserts `options.model === undefined` on endpoint failure.

File: model-lab/correction-intent/eval.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+
+def _gold_labels(rows: list[dict[str, Any]]) -> list[str]:
+    return [str(r.get("label", "")) for r in rows]
+
+
+def _span_overlaps(rows: list[dict[str, Any]], pred_rows: list[dict[str, Any]]) -> list[float]:
+    """Token overlap between predicted + gold correctedAssertion (corrections only)."""
+    overlaps: list[float] = []
+    for gold, pred in zip(rows, pred_rows, strict=True):
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py]: Fixed in b61f427a3. `_span_overlaps` now raises a clear `length mismatch: gold=N pred=M` ValueError before the zip, so the operator sees the descriptive message instead of the cryptic `zip() ... strict=True` one. (`strict=True` dropped since the length is validated explicitly first.) Prove-fail test in `tests/model-lab-integrity-guards.test.mjs`.

File: model-lab/common/manifest_schema.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    if _is_pending(stack.get("python")) and not allow_pending:
+        errors.append("trainingStack.python must record the exact interpreter version")
+
+    libs = stack.get("libs")
+    if not isinstance(libs, Mapping):
+        errors.append("trainingStack.libs must be an object mapping lib → exact version")
+    elif not allow_pending:
+        # A real run must pin at least the load-bearing libs the recipes import.
+        for lib in ("torch", "transformers"):
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/manifest_schema.py]: Fixed in b61f427a3. Strict mode now pins every lib the manifest DECLARES plus the universal `torch`/`transformers`, not just those two — so a correction-intent manifest with `trl`/`peft`/`bitsandbytes` left null fails validation. Declared-set iteration (`dict.fromkeys(("torch","transformers", *libs.keys()))`) respects that the two tasks pin different stacks (encoder vs causal-LM), so a faithfulness run isn't forced to pin causal-LM libs. Prove-fail test asserts a null `trl` is rejected.

File: model-lab/common/manifest_schema.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        # DECLARES. The previous check only covered torch/transformers, so a
+        # correction-intent manifest with trl/peft/bitsandbytes left null passed
+        # strict validation and undermined the no-half-recorded-run gate (kilo
+        # WARNING PRRT_kwDORJXyws6OtyS-). Iterating over declared libs also
+        # respects that the two tasks pin different stacks (encoder vs causal-LM).
+        check_libs = dict.fromkeys(("torch", "transformers", *libs.keys()))
+        for lib in check_libs:
+            ver = libs.get(lib)
+            if _is_pending(ver) or not isinstance(ver, str):
+                errors.append(f"trainingStack.libs.{lib} must pin an exact version for a real run")
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/manifest_schema.py:211]: P2, resolving-with-rationale and folding into the GPU-run follow-up rather than a second push. (This PR's single coherent push already landed b61f427a3 for the HIGHs; each push re-triggers the bot round, and this is completeness polish, not a correctness/security HIGH.)

Agreed the omission case is real, but it needs a task-aware matrix the validator is deliberately WITHOUT: the schema validator is task-agnostic by design (module docstring — "task-agnostic about the inner contract shape"), and it is shared across BOTH tasks. trl/peft are correction-intent-only; accelerate/sentencepiece are faithfulness-gate-only; so trl/peft can't be required unconditionally. A manifest produced by capture_training_stack() records the full installed PINNED_LIBS set from pip freeze, so an omitting manifest can only come from a hand-edit or a broken environment. The strict gate landed in b61f427a3 already requires every DECLARED lib to be pinned (the null-pin case from the sibling thread). Closing the omission case needs a declared per-task required-lib minimum, which belongs in the GPU-run follow-up where the per-task contract is concrete and the real manifests are produced.

Consolidated into follow-up #1700 (item 2) along with the other deferred model-lab nits. Resolving here to keep the PR converging.

File: model-lab/common/manifest_schema.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    # and artifact.hfRepo (where the weights published). Without this a
+    # status:"trained" manifest with eval: {heldOut: null} slips past the
+    # top-level object check and undermines the no-half-recorded-run gate.
+    if not allow_pending:
+        eval_block = manifest.get("eval")
+        if isinstance(eval_block, Mapping) and _is_pending(eval_block.get("heldOut")):
+            errors.append("eval.heldOut is pending — a real run must record held-out metrics")
+        artifact_block = manifest.get("artifact")
+        if isinstance(artifact_block, Mapping) and _is_pending(artifact_block.get("hfRepo")):
+            errors.append("artifact.hfRepo is pending — a real run must publish weights")
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/common/manifest_schema.py:141]: P2, resolving-with-rationale and folding into the GPU-run follow-up rather than a second push (completeness polish on an off-by-default, not-yet-run feature; this PR's single coherent push already landed b61f427a3).

Agreed — strict mode should also reject a pending artifact.revision. It's a one-line addition to the strict artifact block (which already rejects a pending hfRepo as of b61f427a3). Deferring because both hfRepo and revision are only meaningful once common/hf_push.py does a real publish — and hf_push.py is an honestly-labeled STUB until the #1585 GPU-run follow-up (the committed scaffold correctly leaves both null as pending-training). Batching it with the task-aware stack matrix (sibling thread) in the follow-up avoids a second push + bot round for P2 strict-gate completeness.

Consolidated into follow-up #1700 (item 3). Resolving here to keep the PR converging.

File: model-lab/correction-intent/eval.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+def _held_out_is_predictions(held_out: Path, predictions: Path) -> bool:
+    """True when --predictions resolves to the same file as --held-out.
+
+    Pure (no torch) so the guard is unit-testable; ``main`` calls this before
+    loading rows. Uses resolve() so a symlink or a rel/abs alias can't bypass the
+    guard (codex P1 — scoring gold against itself fabricates a perfect eval that
+    could be copied into the manifest, rule 55).
+    """
+    try:
+        return held_out.resolve() == predictions.resolve()
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:129]: Medium, resolving-with-rationale without a third push (off-by-default GPU-lab script; convergence rule + reviewer guidance is to resolve fresh marginal threads without re-pushing).

The guard landed in b61f427a3 prevents the ACCIDENTAL circular-eval case the original P1 flagged — identical path and symlinks (Path.resolve() equality). A hard link is a deliberate circumvention by the operator on their own isolated lab box, outside the guard's threat model: an operator who hard-links the gold file as --predictions to fabricate eval numbers can just hand-edit manifest.json directly (it carries no integrity signature), so inode-checking is marginal hardening against a determined self-faker, not a correctness fix. Adding os.path.samefile / (st_dev, st_ino) equality is a clean 2-line addition — folded into the GPU-run follow-up (#1700) where eval.py gains the real inference loop. Resolving here to keep the PR converging.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:129]: P2, resolving-with-rationale without a third push (same hard-link point as the cursor thread on this line). Path.resolve() catches same-path and symlink circular eval (the original P1); a hard link resolves to a different path but the same inode. That's deliberate self-circumvention on an isolated lab box — the operator who hard-links gold as --predictions to fake metrics can edit manifest.json directly (no integrity signature), so inode equality is marginal hardening, not a correctness gate. os.path.samefile / (st_dev, st_ino) comparison is a clean addition — folded into the GPU-run follow-up (#1700). Resolving to converge.

File: packages/remnic-core/src/extraction-faithfulness.ts [hunk 12]
```diff
@@ -280,13 +284,54 @@ async function callFaithfulnessLlm(
...
+    // the fast path; if it cannot answer within half the budget (min 500ms),
+    // abandon it and let the fallback chain use the remainder (codex P2 — the
+    // advertised graceful fallback must actually reach the chain, not be starved
+    // by a hanging probe sharing the full batch budget).
+    const probeBudgetMs = Math.max(500, Math.floor(timeoutMs / 2));
+    const result = await callOpenAiCompatibleChat(
+      localEndpoint,
+      messages,
+      { temperature: 0.1, maxTokens: 2048, responseFormatJson: true, timeoutMs: probeBudgetMs },
+      fetchImpl,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/extraction-faithfulness.ts:315]: Medium, resolving-with-rationale without a third push — this is the intended design, not a bug. The probe is capped at probeBudgetMs = max(500, floor(timeoutMs/2)), so it finishes before the batch AbortController (signal) fires at batchStart + timeoutMs. The fallback honors that SAME batch signal (callFaithfulnessLlm forwards ...(signal ? { signal } : {}) to the fallback), so total probe + fallback wall time is bounded by the batch budget timeoutMs — NOT probeBudgetMs + timeoutMs. The probeBudgetMs split exists precisely so a wedged probe can't starve the fallback (inline comment lines ~303-309, 'codex P2 — graceful fallback must reach the chain'). A slow-but-successful fallback that hits the batch deadline returns timeout, which IS the graceful-degradation outcome for that batch (subsequent batches use the healthy chain). Forwarding the batch signal into the probe too (instant abort if the batch already elapsed) is a micro-optimization noted in #1700, not a correctness fix.

File: packages/remnic-core/src/extraction-faithfulness.ts [hunk 13]
```diff
@@ -280,13 +284,54 @@ async function callFaithfulnessLlm(
...
+    );
+    if (result?.content) {
+      return { content: result.content, modelUsed: result.modelUsed };
+    }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/extraction-faithfulness.ts:319]: P2, off-by-default, resolving-with-rationale without a third push. When the local endpoint returns 200 with content that isn't valid verdict JSON, the gate surfaces it as malformed_output (facts -> unchecked) rather than silently falling back. That's intentional for an operator who explicitly opted into a local served model: a 200-with-garbage response means the wrong model is served or the prompt regressed, and masking it behind a silent gateway fallback would hide a real misconfiguration. The configured chain remains available for subsequent batches. Pre-validating the endpoint response shape (or making 'fall back on local parse failure' a tunable) is a reasonable resilience enhancement — folded into the GPU-run follow-up (#1700) where the local endpoint contract is concrete. Resolving to converge.
